sound/soc/sof/intel/hda-dsp.c

```diff
@@ -431,11 +436,66 @@ static int hda_dsp_set_D0_state(struct snd_sof_dev *sdev,
 		    !hda_enable_trace_D0I3_S0 ||
 		    sdev->system_suspend_target != SOF_SUSPEND_NONE)
 			flags = HDA_PM_NO_DMA_TRACE;
+
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA)
+		/*
+		 * During system suspend, CORB/RIRB DMA must be stopped and links powered off.
+		 * When the system is in S0 and the iDisp codec is the only HD-Audio codec on
+		 * the platform, CORB/RIRB DMA can be stopped and the iDisp-A link powered off
+		 * when the codec is not powered in order to save power.
+		 */
+		if (sdev->system_suspend_target == SOF_SUSPEND_NONE &&
+		    (bus->codec_powered || bus->codec_mask != HDA_IDISP_CODEC(bus->codec_mask)))
+			goto set_d0i3_reg;
+		else
+			hda->idsip_link_off_in_S0 = true;
+
+		/* stop the CORB/RIRB DMA if it is On */
+		if (bus->cmd_dma_state)
+			snd_hdac_bus_stop_cmd_io(bus);
+
+		/* power off links */
+		ret = snd_hdac_ext_bus_link_power_down_all(bus);
+		if (ret < 0) {
+			dev_err(sdev->dev, "error: failed to power down links %d\n", ret);
+			return ret;
+		}
+#endif
 	} else {
 		/* prevent power gating in D0I0 */
 		flags = HDA_PM_PPG;
+
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA)
+		/*
+		 * During system resume, CORB/RIRB DMA and links must be brought to the same state
+		 * as before suspend. But when the system is in S0, this should be done if the
+		 * iDisp link was powered off when entering D0i3. The codec_powered status
+		 * cannot be used as it could have changed when the DSP was in D0i3.
+		 */
+		if (sdev->system_suspend_target == SOF_SUSPEND_NONE && !hda->idsip_link_off_in_S0)
+			goto set_d0i3_reg;
+		else
+			hda->idsip_link_off_in_S0 = false;
```

Comment:
ditto for the inversion and the `else`

Reply:
@lyakh @kv2019i does it look better now?